pkg/apis/core/v1/defaults.go

```diff
@@ -349,6 +349,23 @@ func SetDefaults_HTTPGetAction(obj *v1.HTTPGetAction) {
 		obj.Scheme = v1.URISchemeHTTP
 	}
 }
+
+// SetDefaults_Namespace adds a default label for all namespaces
+func SetDefaults_Namespace(obj *v1.Namespace) {
+	// TODO, remove this in 1.22
```

Comment:
wrap in a check to see if 1) the namespace has a name and 2) the label does not match it:

```go
if len(obj.Name) > 0 && obj.Labels[v1.LabelMetadataName] != obj.Name {
  if utilfeature.DefaultFeatureGate.Enabled(features.NamespaceDefaultLabelName) {
    ... check for nil labels ...
    ... assign the label ...
  }
}
```

a namespace being created with a `generateName` will not have a name yet when it hits defaulting, and a namespace with a label that already matches doesn't need to take a read lock on the feature gate

Reply:
add a test case creating a namespace without a name set and with `GenerateName: "myname-"` set to make sure the namespace returned from the create API call has the label and the namespace returned from the next `get` API call has the label

Reply:
hmmm Where would a `Get` testcase live ? defaults_test seems to only test roundtrip serialization, and doesnt make any API calls to a mock server?

Reply:
ahhh validation_test i think

Reply:
` make test-integration WHAT=./test/integration/namespace GOFLAGS="-v"` <-- putting it in here